File: 5-Problem-A-Day/59. Spiral Matrix II.java
class Solution {
    public int[][] generateMatrix(int n) {
        int[][] matrix = new int[n][n];
        int start = 1;
        for (int i = 0; i <= (n - 1) / 2; i++) {
            start = fill(matrix, i, i, n - i - 1, n - i - 1, start);
        }
        return matrix;
    }

    private int fill(int[][] matrix, int r1, int c1, int r2, int c2, int start) {
        for (int c = c1; c <= c2; c++) {
            matrix[r1][c] = start;
            start += 1;
        }
        for (int r = r1 + 1; r <= r2; r++) {
            matrix[r][c2] = start;
            start += 1;
        }
        for (int c = c2 - 1; c >= c1; c--) {
            matrix[r2][c] = start;
            start += 1;
        }
        for (int r = r2 - 1; r >= r1 + 1; r--) {
            matrix[r][c1] = start;
            start += 1;
        }
        return start;
    }
}

File: 2021/59. Spiral Matrix II.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> matrix(n, vector<int>(n, 1));
        int num = 1;
        for (int i = 0; i < (n + 1) / 2; i++) {
            for (int j = i; j < n - i; j++) {
                matrix[i][j] = num;
                num++;
            }
            for (int j = i + 1; j < n - i; j++) {
                matrix[j][n - 1 - i] = num;
                num++;
            }
            for (int j = n - 2 - i; j >= i; j--) {
                matrix[n - 1 - i][j] = num;
                num++;
            }
            for (int j = n - 2 - i; j > i; j--) {
                matrix[j][i] = num;
                num++;
            }
        }
        return matrix;
    }
};